query-algebrizer/src/clauses/or.rs

```diff
@@ -658,6 +658,21 @@ impl ConjoiningClauses {
             type_associations = type_needed.iter().cloned().collect();
         }
 
+        // Collect the new type information from the arms. There's some redundant work here --
+        // they already have all of the information from the parent.
```

Comment:
The fact that we push type information down is strictly an optimization, right?  That is, this problem decomposes as a post-order tree walk followed by a type narrowing union or intersection.  We can cull parts of the tree by doing the intersection inline if we push type information into the children.